src/main/java/edu/cornell/kfs/sys/service/impl/AwsSecretServiceImpl.java

```diff
@@ -33,9 +36,72 @@
     protected String kfsSharedNamespace;
     protected int retryCount;
     
+    private Map<String, String> awsSecretsCache;
+    
+    @Override
+    public void clearCache() {
+        awsSecretsCache = new ConcurrentHashMap<>();
+    }
+    
+    @Override
+    public void logCacheStatus() {
+        if (LOG.isDebugEnabled()) {
+            if (awsSecretsCache == null) {
+                LOG.debug("logCacheStatus, the cache has not been set");
+            } else if (awsSecretsCache.isEmpty()) {
+                LOG.debug("logCacheStatus, the cache has been set, but is empty");
+            } else {
+                for (String key : awsSecretsCache.keySet()) {
+                    LOG.debug("logCacheStatus, AWS secret key: " + key);
+                }
+            }
+        }
+    }
+    
     @Override
     public String getSingleStringValueFromAwsSecret(String awsKeyName, boolean useKfsInstanceNamespace) {
+        createCacheIfNotPresent();
         String fullAwsKey = buildFullAwsKeyName(awsKeyName, useKfsInstanceNamespace);
+        String cachedSecretValue = retrieveSecretFromCache(fullAwsKey);
+        
+        if (StringUtils.isNotBlank(cachedSecretValue)) {
+            return cachedSecretValue;
+        } else {
+            LOG.debug("getSingleStringValueFromAwsSecret, in cache there was no value for " + fullAwsKey);
+            String secretValue = retrieveSecretFromAws(fullAwsKey);
+            updateCacheValue(fullAwsKey, secretValue);
+            return secretValue;
+        }
+    }
+
+    private void createCacheIfNotPresent() {
+        if (awsSecretsCache == null) {
+            LOG.debug("createCacheIfNotPresent, Cache has not been setup, so run the clear cache to create a new cache element");
+            clearCache();
+        }
+    }
+    
+    private String retrieveSecretFromCache(String fullAwsKey) {
+        if (awsSecretsCache.containsKey(fullAwsKey)) {
```

Comment:
On a ConcurrentHashMap, a null key will cause the containsKey() and get() methods to throw an exception. You may want to check that the key is non-null before attempting to access the Map. Sorry for not clarifying that sooner.

Reply:
in practice, if a null secret name gets passed in, it will get translated to "kfs/local-dev/null" or something similar depending on the namespace for the environment.
However, I added some logic and throw a meaningful exception when this happens